builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,342 @@
+
+func (lm *lockManager) GetPolicyUpsert(storage logical.Storage, name string, derived bool) (*Policy, bool, bool, error) {
+	return lm.getPolicyCommon(storage, name, true, derived)
+}
+
+// When the function returns, a lock will be held on the policy if err == nil.
+// The type of lock will be indicated by the return value. It is the caller's
+// responsibility to unlock.
+func (lm *lockManager) getPolicyCommon(storage logical.Storage, name string, upsert, derived bool) (p *Policy, lockType bool, upserted bool, err error) {
+	// If we are using a cache, lock it now to avoid having to do really
+	// complicated lock juggling as we call various functions. We'll also defer
+	// the store into the cache.
+	lockType = shared
+	lm.lockPolicy(name, shared)
+
+	if lm.CacheActive() {
+		lm.globalMutex.RLock()
+		p = lm.cache[name]
+		if p != nil {
+			lm.globalMutex.RUnlock()
+			return
+		}
+		lm.globalMutex.RUnlock()
+
+		// When we return, since we didn't have the policy in the cache, if
+		// there was no error, write the value in.
+		defer func() {
+			lm.globalMutex.Lock()
+			defer lm.globalMutex.Unlock()
+			// Make sure a policy didn't appear. If so, it will only be set if
+			// there was no error, so now just clear the error and return that
+			// policy.
+			exp := lm.cache[name]
+			if exp != nil {
+				upserted = false
+				err = nil
+				p = exp
+				return
+			}
+
+			if err == nil {
+				lm.cache[name] = p
+			}
+		}()
+	}
+
+	p, err = lm.getStoredPolicy(storage, name)
+	if err != nil {
+		lm.UnlockPolicy(name, shared)
+		return
+	}
+
+	if p == nil {
+		if !upsert {
+			lm.UnlockPolicy(name, shared)
+			return
+		}
+
+		// Get an exlusive lock; on success, check again to ensure that no
+		// policy exists. Note that if we are using a cache we will already be
+		// serializing this entire code path and it's currently the only one
+		// that generates policies, so we don't need to check the cache here;
+		// simply checking the disk again is sufficient.
+		lm.UnlockPolicy(name, shared)
+		lockType = exclusive
+		lm.lockPolicy(name, exclusive)
+
+		p, err = lm.getStoredPolicy(storage, name)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			return
+		}
+		if p != nil {
+			return
+		}
+
+		upserted = true
+
+		p = &Policy{
+			Name:       name,
+			CipherMode: "aes-gcm",
+			Derived:    derived,
+		}
+		if derived {
+			p.KDFMode = kdfMode
+		}
+
+		err = p.rotate(storage)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			p = nil
+		}
+
+		// We don't need to worry about upgrading since it will be a new policy
+		return
+	}
+
+	if p.needsUpgrade() {
+		lm.UnlockPolicy(name, shared)
+		lockType = exclusive
+		lm.lockPolicy(name, exclusive)
+
+		// Reload the policy with the write lock to ensure we still need the upgrade
+		p, err = lm.getStoredPolicy(storage, name)
+		if err != nil {
+			lm.UnlockPolicy(name, exclusive)
+			return
+		}
+		if p == nil {
+			err = fmt.Errorf("error reloading policy for upgrade")
```

Comment:
This shouldn't be an error, it should just look like an operation on a key that doesn't exist (e.g. p == nil). Since the operation ordering is not a guarantee, this means delete just got ordered first.